### PR TITLE
[ROCm][Perf] DSv3.2: fuse indexer Q-RoPE+quant + K-norm/RoPE/quant/cache

### PR DESCRIPTION
**Pull request by frida-andersson**
Port [ROCm/ATOM#788](https://github.com/ROCm/ATOM/pull/788)'s indexer fused-kernel shape into vLLM. Replaces the `k_norm → rotary_emb → per_token_group_quant_fp8 → cache` chain (and the Q-side `per_token_group_quant_fp8`) inside `Indexer.forward` with a single AITER kernel `aiter.indexer_qk_rope_quant_and_cache` (ROCm/aiter#3185, correctness fixes in ROCm/aiter#3451).

> **Companion:** ROCm/aiter#3451 — required. Without it GSM8K strict-match drops 0.95 → 0.49.

## Gating

`VLLM_ROCM_USE_AITER_DSV32_INDEXER_FUSION` (default off) + main AITER toggle; auto-disabled when the running AITER build does not expose `indexer_qk_rope_quant_and_cache`. The kernel hard-asserts `index_head_dim == 128`, `qk_rope_head_dim == 64`, `head_dim == quant_block_size`; `Indexer.__init__` only enables the path if all three hold and `model_type != "glm_moe_dsa"`.

## Design notes

- **Stays inside the existing custom op.** Calling the fused kernel from inside `rocm_aiter_sparse_attn_indexer` keeps `q_fp8` and `weights_out` from escaping — an earlier version that registered them as a separate op leaked pointers across piecewise-cudagraph segments and tanked GSM8K to ~1%.
- **Persistent workspace** for `q_fp8` / `weights_out`, grown monotonically. Otherwise 61 indexer layers × 2 allocations per decode step churn the eager allocator.
- **K stride fixup.** The kernel asserts `k.stride(0) == head_dim`, but the upstream `.contiguous()` is a no-op on `size(0) == 1` views. We rewrite the metadata via `as_strided` when it's safe, else `empty + copy_`.
- **`preshuffle=True`.** Downstream consumers read MFMA 16×16 SHUFFLE; the original call site passed `False` and silently produced garbage. Bytewise-equal to the unfused triton write after the fix.
- **fp32 LN params** passed through as-is (required by ROCm/aiter#3451).

## Correctness (GSM8K 20-shot strict-match, 1319 questions, `num_concurrent=256`)

| Configuration | strict-match |
|---|---|
| Fusion off, fp8 KV cache (baseline) | 0.95 ± 0.006 |
| Fusion on, **without** ROCm/aiter#3451 + `preshuffle=True`, fp8 | 0.49 |
| Fusion on, with ROCm/aiter#3451 + `preshuffle=True`, fp8 | **0.95 ± 0.006** |
| Fusion on, with ROCm/aiter#3451 + `preshuffle=True`, bf16 | **0.9591 ± 0.0055** |

## Performance (MI355X, DSv3.2, TP=4, 1k/100/mc=4)

| Metric | OFF | ON | Δ |
|---|---|---|---|
| Median ITL (FP8) | 14.01 ms | 13.71 ms | −2.1% |
| Output throughput (FP8) | 217.4 tok/s | 224.5 tok/s | +3.3% |
| Per-step CUDA (FP8) | 16.880 ms | 16.421 ms | −0.459 ms (−2.72%) |
| Per-step CUDA (BF16) | 16.866 ms | 16.255 ms | −0.611 ms (−3.62%) |

Four indexer kernels (K-norm, the RoPE helper, the K-cache write, and one half of the per-token-group quant) collapse into a single fused launch — net −483.6 ms across 776 decode steps, or about 0.62 ms per step.

Before:
before_index_fusion
After:
after_index_fusion

## References

- Parent fusion: ROCm/ATOM#788
- AITER kernel: ROCm/aiter#3185
- Companion AITER fixes: ROCm/aiter#3451

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -122,6 +122,8 @@
     VLLM_ROCM_USE_AITER_MHA: bool = True
     VLLM_ROCM_USE_AITER_FP4_ASM_GEMM: bool = False
     VLLM_ROCM_USE_AITER_TRITON_ROPE: bool = False
+    VLLM_ROCM_USE_AITER_DSV32_INDEXER_FUSION: bool = False
+    VLLM_ROCM_DSV32_INDEXER_FUSION_MAX_LAYERS: int = -1
```
Comment (AndreasKaratzas) [inline vllm/envs.py]: Please find an alternative to env vars.

File: vllm/model_executor/models/deepseek_v2.py [hunk 2]
```diff
@@ -663,6 +663,41 @@ def __init__(
         from vllm.v1.attention.backends.mla.indexer import get_max_prefill_buffer_size
 
         self.max_total_seq_len = get_max_prefill_buffer_size(vllm_config)
+        # When the fused kernel runs it writes K into the indexer cache itself,
+        # so SparseAttnIndexer must skip its own indexer_k_quant_and_cache call.
+        # The kernel hard-asserts the DSv3.2 shape (head_dim 128, rope_dim 64)
+        # and is not registered for the GLM 5 DSA variant.
```
Comment (AndreasKaratzas) [inline vllm/model_executor/models/deepseek_v2.py]: Shorten this comment.

File: vllm/model_executor/models/deepseek_v2.py [hunk 3]
```diff
@@ -663,6 +663,41 @@ def __init__(
...
+                        break
+                if layer_idx < 0 or layer_idx >= max_fused_layers:
+                    self.use_qk_rope_cache_fusion = False
+        # Folded into weights_out by the fused kernel alongside Q's per-token
+        # scale; the unfused path applies it after Q quant.
+        self._weights_scale = self.softmax_scale * self.n_head**-0.5
+        # The fused kernel asserts norm_{weight,bias}.dtype == q.dtype, but
+        # LayerNorm stores both as fp32. Cast once on first forward (lands
+        # inside profile_run, pre-cudagraph-capture) and reuse so the
+        # captured graph sees a stable address.
```
Comment (AndreasKaratzas) [inline vllm/model_executor/models/deepseek_v2.py]: I think this can go

**Review by AndreasKaratzas — COMMENTED** [review]
cc @dllehr-amd @Rohan138

File: vllm/envs.py (hunk 1 above)
Comment (frida-andersson) [reply vllm/envs.py]: Thanks @AndreasKaratzas for the review. `VLLM_ROCM_DSV32_INDEXER_FUSION_MAX_LAYERS` is a debug bisection knob, happy to drop.`VLLM_ROCM_USE_AITER_DSV32_INDEXER_FUSION` is the on/off gate. Capability gating is already covered by the lazy `hasattr(aiter, "indexer_qk_rope_quant_and_cache")` probe - the flag's only remaining job is a runtime kill switch / bench A-B, like the other `VLLM_ROCM_USE_AITER_*` flags.

Two options:
1. Drop it, auto-enable when AITER + kernel are present (disable via `VLLM_ROCM_USE_AITER=0`).
2. Keep it, flip default to on.

Lean toward (1), but happy with (2) if you'd rather keep a targeted disable. Let me know what you think

**Comment by frida-andersson** [comment]
Update — accuracy regression found and fixed.

After landing the original fusion plumbing in this PR, GSM8K 20-shot strict-match dropped from 0.95 (fusion off) to 0.49 (fusion on) on DSv3.2 / MI355X. Root-caused to three independent bugs:

1. **AITER kernel — Q quant** skipped UE8M0 rounding and used direct division by `q_fp8_max`, while the unfused vLLM `per_token_group_quant_fp8` path multiplies by the reciprocal and applies UE8M0. Disagreed with the unfused path on ~94% of Q values per call.
2. **AITER kernel — `norm_weight` / `norm_bias`** were typed `scalar_t*` (= bf16 at runtime), so the fp32 `LayerNorm.weight` / `bias` got cast down before launch. Lost ~16 mantissa bits, drifted K on ~3% of values per call.
3. **vLLM call site — `preshuffle=False`**. The downstream gather and `paged_mqa_logits` consumers read the indexer cache in MFMA 16×16 SHUFFLE layout, but the fused kernel was writing NORMAL. The kernel already supports both via the `preshuffle` arg — we were just passing the wrong value.

Fixes 1 and 2 are kernel-side and live in the companion AITER PR ROCm/aiter#3451. Fix 3 plus the call-site change to pass fp32 LN params (now required by ROCm/aiter#3451's tightened dtype assert) is what the latest commits in this PR do.

After all three: GSM8K 20-shot strict-match recovers to 0.95 ± 0.006 (fp8 KV cache) and 0.9591 ± 0.0055 (bf16 KV cache) — matches the unfused baseline.

File: vllm/envs.py (hunk 1 above)
Comment (tjtanaa) [reply vllm/envs.py]: Let's go with option 1 since if it is going to be the best solution for all cases.

File: vllm/model_executor/layers/sparse_attn_indexer.py
Comment (tjtanaa) [inline vllm/model_executor/layers/sparse_attn_indexer.py:471]: We will also need to update the signature for `forward_cuda` and other non-hip forward pass

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 4]
```diff
@@ -23,6 +23,49 @@
     _ON_GFX950 = False
 
 
+# Module-level workspace for the fused indexer outputs. Per-call torch.empty
+# returned addresses that differed between cudagraph capture and replay,
+# which the downstream rocm_fp8_*_mqa_logits then read as garbage.
+_FUSED_INDEXER_Q_FP8_BUF: torch.Tensor | None = None
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Maybe try using `from vllm.v1.worker.workspace import current_workspace_manager`

Comment (frida-andersson) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Thanks for the recommendation - replaced the module-level workspace globals with `current_workspace_manager().get_simultaneous()`

File: vllm/model_executor/layers/sparse_attn_indexer.py
Comment (frida-andersson) [reply vllm/model_executor/layers/sparse_attn_indexer.py:471]: Added `**fusion_kwargs` to forward_cuda/forward_xpu for signature consistency

File: vllm/envs.py (hunk 1 above)
Comment (frida-andersson) [reply vllm/envs.py]: Dropped both `VLLM_ROCM_USE_AITER_DSV32_INDEXER_FUSION` and the `MAX_LAYERS` debug knob. Fusion now auto-enables when `VLLM_ROCM_USE_AITER=1` and the kernel is present (lazy hasattr probe unchanged).

**Comment by frida-andersson** [comment]
Thanks for the feedback @tjtanaa and @AndreasKaratzas! Summary of changes:

- Dropped `VLLM_ROCM_USE_AITER_DSV32_INDEXER_FUSION` and `VLLM_ROCM_DSV32_INDEXER_FUSION_MAX_LAYERS`. Fusion now auto-enables when `VLLM_ROCM_USE_AITER=1` and the running AITER build exposes indexer_qk_rope_quant_and_cache (lazy hasattr probe)
- Added `**fusion_kwargs` to forward_cuda and forward_xpu for signature consistency
- Replaced the module-level workspace globals with `current_workspace_manager().get_simultaneous()`

Re-verified after these changes: GSM8K 20-shot strict-match 0.9469 (fp8 KV), profiler confirms indexer_qk_rope_quant_and_cache firing for all 61 layers.

**Note**: this PR depends on **[ROCm/aiter#3451](https://github.com/ROCm/aiter/pull/3451)**, without it GSM8K 20-shot drops to 0.49. Is there a preferred way to express that dependency here?

**Comment by frida-andersson** [comment]
**Update**: since this PR depends on https://github.com/ROCm/aiter/pull/3451 and the aiter bump, I'll wait to rebase it until those are merged

**Comment by maeehart** [comment]
GitHub reports CONFLICTING. Can you rebase onto current main and push?

**Comment by frida-andersson** [comment]
Rebased onto current `main`. Fusion still auto-enables when `VLLM_ROCM_USE_AITER=1` and AITER has `indexer_qk_rope_quant_and_cache`.

One extra hook: no-op `record_logical_topk_ready` so `--performance-mode throughput` can compile. Fused decode logits use `torch.full` so workspace offset 0 does not alias `q_fp8`.

Verified on 8×MI350, TP4, DSv3.2:

| conc | off median ITL | on median ITL | ITL Δ | tok/s Δ |
|---|---|---|---|---|
| 1 | 13.750 ms | 13.370 ms | −2.76% | +2.57% |
| 4 | 15.990 ms | 15.583 ms | −2.54% | +1.85% |
| 8 | 18.047 ms | 17.820 ms | −1.26% | +0.51% |

GSM8K 20-shot 0.96, fused vs unfused `q_fp8` + K-cache is byte-exact with `preshuffle=True`

https://github.com/ROCm/aiter/pull/3451 is now merged with corresponding aiter bump, PTAL @AndreasKaratzas @tjtanaa

File: vllm/_aiter_ops.py [hunk 5]
```diff
@@ -2246,6 +2247,22 @@ def is_triton_rotary_embed_enabled(cls) -> bool:
...
+    @if_aiter_supported
+    def is_dsv32_indexer_qk_fusion_enabled(cls) -> bool:
+        if not cls._AITER_ENABLED:
+            return False
+        if cls._DSV32_INDEXER_QK_FUSION_KERNEL_PRESENT is None:
+            try:
+                import aiter
+
+                cls._DSV32_INDEXER_QK_FUSION_KERNEL_PRESENT = hasattr(
+                    aiter, "indexer_qk_rope_quant_and_cache"
```
Comment (simondanielsson) [inline vllm/_aiter_ops.py]: **Suggestion**: can we drop this check and attribute now given indexer_qk_rope_quant_and_cache is shipped with aiter and instead just gate on aiter being enabled?

**Review by simondanielsson — COMMENTED** [review]
I see https://github.com/vllm-project/vllm/pull/51315 also covers this albeit with a slightly different implementation (I'm aware your PR was first). Did you also measure a similar prefill/ttft regression when enabling this fusion?

**Comment by frida-andersson** [comment]
Thanks for the comment @simondanielsson! I re-measured this now on gfx942, DSv3.2, TP4 and couldn't see any #51315-style prefill/TTFT hit: 8k/1k median TTFT 845 → 844 ms, 131k/1k 24.09 → 24.14 s.

Decode does improve on the original 1k/100 mc=4 recipe: median ITL ≈15.61 → ≈15.23 ms (−2.4%), throughput ≈197 → ≈201 tok/s, rank-0 decode CUDA 9.55 → 9.21 ms/step (−3.6%).
